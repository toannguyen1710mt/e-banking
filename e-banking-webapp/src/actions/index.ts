export * from './auth';
export * from './transaction';
export * from './account';
