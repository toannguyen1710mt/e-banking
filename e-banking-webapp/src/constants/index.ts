export * from './accounts';
export * from './routes';
export * from './themes';
export * from './messages';
export * from './images';
export * from './regex';
export * from './navbar';
export * from './titles';
export * from './global-transfer';
export * from './balance';
export * from './charts';
export * from './form';
export * from './api';
export * from './tags';
export * from './settings';
export * from './tables';
export * from './cards';
export * from './date';
export * from './sidebar-list';
