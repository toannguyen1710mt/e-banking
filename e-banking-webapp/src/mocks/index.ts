export * from './options';
export * from './card';
export * from './charts';
export * from './transaction';
export * from './table';
export * from './transferConfirmData';
export * from './target';
export * from './analytics';
export * from './auth';
export * from './ServiceCard';
export * from './formMocks';
