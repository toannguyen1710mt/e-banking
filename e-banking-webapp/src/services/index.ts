export * from './http-client';
export * from './account';
export * from './transaction';
export * from './users';
