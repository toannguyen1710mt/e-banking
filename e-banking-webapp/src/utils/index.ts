export * from './number';
export * from './calculator';
export * from './chart';
export * from './date';
export * from './convertToUSD';
export * from './query-params';
export * from './futureMonth';
export * from './validation';
export * from './search-params';
export * from './sanitizeAmount';
export * from './toastManager';
