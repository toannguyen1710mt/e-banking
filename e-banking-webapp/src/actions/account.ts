'use server';

// Libs
import { revalidateTag } from 'next/cache';

// Constants
import { API_ENDPOINTS, ERROR_MESSAGES } from '@/constants';

// Interfaces
import { IAccountPayloadData } from '@/interfaces';

// Services
import { httpClient } from '@/services';

export const updateAccountInfo = async (
  accountDocumentId: string,
  accountData: IAccountPayloadData,
) => {
  try {
    await httpClient.put(`${API_ENDPOINTS.ACCOUNTS}/${accountDocumentId}`, {
      data: accountData,
    });

    revalidateTag(API_ENDPOINTS.ACCOUNTS);
  } catch (error) {
    throw new Error(ERROR_MESSAGES.ERROR_UPDATE_ACCOUNT_INFO + error);
  }
};
