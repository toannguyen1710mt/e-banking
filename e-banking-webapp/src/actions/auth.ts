'use server';

import {
  signIn,
  signOut as nextAuthSignOut,
  unstable_update,
} from '@/config/auth';
import { revalidateTag } from 'next/cache';
import { AuthError } from 'next-auth';

// Constants
import { ERROR_MESSAGES, API_ENDPOINTS, TAGS } from '@/constants';

// Services
import { httpClient } from '@/services/http-client';

// Interfaces
import {
  TSignUpPayload,
  TSignInFormData,
  TUpdateInfo,
  AuthResponse,
  IAccount,
  IAccountPayload,
  ICardPayload,
  SuccessResponse,
  TChangePasswordFormData,
  TChangePasswordSuccessResponse,
  IUser,
  ICard,
} from '@/interfaces';

export const authenticateUser = async (formData: TSignInFormData) => {
  try {
    await signIn('credentials', formData);
  } catch (error) {
    if (error instanceof AuthError) {
      return ERROR_MESSAGES.ACCOUNT_AND_PASSWORD_INVALID;
    }
    throw error;
  }
};

export const signUp = async (signUpData: TSignUpPayload) => {
  const response = await httpClient.post<AuthResponse>(
    API_ENDPOINTS.SIGN_UP,
    signUpData,
  );

  return response;
};

export const signOut = async () => await nextAuthSignOut({ redirect: false });

export const changePassword = async (
  payload: TChangePasswordFormData,
  jwt: string,
) => {
  const response = await httpClient.post<TChangePasswordSuccessResponse>(
    API_ENDPOINTS.CHANGE_PASSWORD,
    payload,
    {
      headers: {
        Authorization: `Bearer ${jwt}`,
      },
    },
  );

  return response;
};

export const updateUser = async (id: number, payload: TUpdateInfo) => {
  try {
    const response = await httpClient.put(
      `${API_ENDPOINTS.USERS}/${id}`,
      payload,
    );

    await unstable_update({ user: response.data });

    return response.data;
  } catch (error) {
    if (error instanceof AuthError) {
      throw ERROR_MESSAGES.SIGN_UP_ERROR;
    }
  }
};

export const addAccount = async (payload: IAccountPayload) => {
  try {
    const response = await httpClient.post<SuccessResponse<IAccount>>(
      API_ENDPOINTS.ACCOUNTS,
      payload,
    );

    return response.data;
  } catch (error) {
    if (error instanceof AuthError) {
      throw ERROR_MESSAGES.SIGN_UP_ERROR;
    }
  }
};

export const addCard = async (payload: ICardPayload) => {
  try {
    const response = await httpClient.post(API_ENDPOINTS.CARDS, payload);

    return response.data;
  } catch (error) {
    if (error instanceof AuthError) {
      throw ERROR_MESSAGES.SIGN_UP_ERROR;
    }
  }
};

export const updateEmailSettings = async (
  userId: number,
  payload: Partial<IUser>,
) => {
  await httpClient.put(`${API_ENDPOINTS.USERS}/${userId}`, payload);

  return revalidateTag(TAGS.USERS);
};

export const getUser = async (id: number): Promise<IUser | undefined> => {
  try {
    const response = await httpClient.get(`${API_ENDPOINTS.USERS}/${id}`);

    return response.data;
  } catch (error) {
    if (error instanceof AuthError) {
      throw ERROR_MESSAGES.SIGN_UP_ERROR;
    }
  }
};

export const addNewCardByAccountId = async (
  accountId: string,
  payload: Omit<ICard, 'id' | 'fullName' | 'account'>,
) => {
  try {
    const response = await httpClient.post(API_ENDPOINTS.CARDS, {
      data: {
        ...payload,
        account: {
          connect: [accountId],
        },
      },
    });

    revalidateTag(TAGS.CARD);

    return response.data;
  } catch (error) {
    if (error instanceof AuthError) {
      throw ERROR_MESSAGES.ADD_CARD_FAILED;
    }
  }
};
