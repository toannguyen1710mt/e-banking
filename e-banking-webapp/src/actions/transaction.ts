'use server';

// Libs
import { revalidateTag } from 'next/cache';

// Constants
import { API_ENDPOINTS, ERROR_MESSAGES } from '@/constants';

// Interfaces
import { ITransaction, TransactionCreateData } from '@/interfaces';

// Services
import { httpClient } from '@/services';

export const createTransaction = async (
  accountId: string,
  transactionData: TransactionCreateData,
) => {
  try {
    await httpClient.post<ITransaction>(API_ENDPOINTS.TRANSACTIONS, {
      data: {
        ...transactionData,
        account: {
          connect: [accountId],
        },
      },
    });

    revalidateTag(API_ENDPOINTS.TRANSACTIONS);
  } catch (error) {
    throw new Error(ERROR_MESSAGES.ERROR_CREATE_TRANSACTION + error);
  }
};
