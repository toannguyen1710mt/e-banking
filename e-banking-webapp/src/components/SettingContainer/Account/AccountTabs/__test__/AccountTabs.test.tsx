import { act, fireEvent, render, waitFor } from '@testing-library/react';

// Services
import { getUserById } from '@/services';

// Actions
import { updateEmailSettings } from '@/actions';

// Mocks
import { MOCK_SESSION_DATA } from '@/mocks';

// Utils
import { toastManager } from '@/utils';

// Components
import { AccountTabs } from '..';
import { ERROR_MESSAGES } from '@/constants';

jest.mock('@/services', () => ({
  getUserById: jest.fn(),
}));

jest.mock('@/actions', () => ({
  updateEmailSettings: jest.fn(),
}));

jest.mock('@/utils', () => ({
  ...jest.requireActual('@/utils'),
  toastManager: {
    showToast: jest.fn(),
  },
}));

describe('AccountTabs component', () => {
  beforeEach(() => {
    global.confirm = jest.fn();

    (getUserById as jest.Mock).mockResolvedValue({
      user: {
        announcements: true,
        updates: true,
        feedbacksAndSurvey: true,
        events: true,
        generalNotification: true,
        promotions: true,
        eventsNearMe: true,
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Should render snapshot correctly.', async () => {
    await act(async () => {
      const { container } = render(<AccountTabs session={MOCK_SESSION_DATA} />);
      expect(container).toMatchSnapshot();
    });
  });

  it('Should log an error when fetching user data fails', async () => {
    (getUserById as jest.Mock).mockRejectedValue(new Error('Fetch error'));

    render(<AccountTabs session={MOCK_SESSION_DATA} />);

    await waitFor(() => {
      expect(toastManager.showToast).toHaveBeenCalledWith(
        expect.stringContaining(ERROR_MESSAGES.ERROR_FETCHING_DATA),
        'error',
        'top-center',
      );
    });
  });

  it('Should update email settings successfully', async () => {
    const { getByLabelText, getByTestId, getAllByLabelText } = render(
      <AccountTabs session={MOCK_SESSION_DATA} />,
    );

    fireEvent.click(getByLabelText('label-account-tabs-1'));

    await waitFor(() => {
      const checkboxes = getAllByLabelText('check-references-0');
      fireEvent.click(checkboxes[0]);

      fireEvent.click(getByTestId('update-email-preferences'));
    });

    expect(updateEmailSettings).toHaveBeenCalledWith(
      82,
      expect.objectContaining({
        announcements: true,
        events: false,
        eventsNearMe: false,
        feedbacksAndSurvey: false,
        generalNotification: false,
        promotions: false,
        updates: false,
      }),
    );
  });

  it('Should return an error message when update fails', async () => {
    const errorMessage = 'Failed to update email settings';
    (updateEmailSettings as jest.Mock).mockRejectedValue(
      new Error(errorMessage),
    );
    (getUserById as jest.Mock).mockResolvedValue({
      user: {
        announcements: false,
        updates: false,
        feedbacksAndSurvey: false,
        events: false,
        generalNotification: false,
        promotions: false,
        eventsNearMe: false,
      },
    });

    const { getByLabelText, getByTestId, getAllByLabelText, getByText } =
      render(<AccountTabs session={MOCK_SESSION_DATA} />);

    fireEvent.click(getByLabelText('label-account-tabs-1'));

    await waitFor(() => {
      const checkboxes = getAllByLabelText('check-references-0');
      fireEvent.click(checkboxes[0]);

      fireEvent.click(getByTestId('update-email-preferences'));
    });

    expect(getByText('Email Settings')).toBeInTheDocument();
  });

  it('Should handle confirm navigation', async () => {
    jest.spyOn(window, 'confirm').mockReturnValue(true);
    const { getByLabelText } = render(
      <AccountTabs session={MOCK_SESSION_DATA} />,
    );

    fireEvent.change(getByLabelText('current password'), {
      target: { value: '123' },
    });

    fireEvent.click(getByLabelText('label-account-tabs-1'));
    expect(global.confirm).toHaveBeenCalledWith(
      'The entered data will not be saved. Do you want to leave?',
    );
  });
});
