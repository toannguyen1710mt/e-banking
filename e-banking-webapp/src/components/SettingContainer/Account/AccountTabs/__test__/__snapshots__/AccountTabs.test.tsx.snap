// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountTabs component Should render snapshot correctly. 1`] = `<div />`;
